python/sglang/srt/managers/multimodal_processors/base_processor.py

```diff
@@ -370,29 +368,102 @@ def get_mm_items_offset_by_pair(
+        # 1. Check each input and categorize
+        # ------------------------------------------------------------
+        has_image = False
+        has_pixel_values = False
+        has_precomputed_features = False
+
+        for mm_input in mm_inputs:
+            if isinstance(mm_input, Image.Image):
+                has_image = True
+            elif isinstance(mm_input, dict):
+                if mm_input.get("precomputed_features", None) is not None:
+                    has_precomputed_features = True
+                elif mm_input.get("pixel_values", None) is not None:
+                    has_pixel_values = True
+                else:
+                    raise ValueError(
+                        f"Invalid multimodal input: {mm_input}, expected dict with pixel_values or precomputed_features"
+                    )
+            else:
+                raise ValueError(
+                    f"Invalid multimodal input: {mm_input}, expected Image.Image or dict"
+                )
+
+        # Count how many different formats we found
+        format_count = sum([has_image, has_pixel_values, has_precomputed_features])
+
+        if format_count > 1:
             raise ValueError(
-                "Unsupported: mixture of multimodal inputs where some but not all are preprocessed."
+                "Unsupported: mixture of multimodal input formats. "
+                f"Found formats: image={has_image}, pixel_values={has_pixel_values}, "
+                f"precomputed_features={has_precomputed_features}"
             )
-        return ret
+
+        if has_image:
+            return None
+
+        # ------------------------------------------------------------
+        # 2. Combine the inputs into a single MultimodalDataItem
+        # ------------------------------------------------------------
+
+        if has_precomputed_features:
+            combined_mm_item = MultimodalDataItem(modality=Modality.IMAGE)
+            combined_mm_item.precomputed_features = self._extract_processor_features(
+                mm_inputs, "precomputed_features"
+            )
+        elif has_pixel_values:
+            values = self._extract_processor_features_from_all_attributes(mm_inputs)
+            combined_mm_item = MultimodalDataItem.from_dict(values)
+        else:
+            raise ValueError(f"Invalid multimodal inputs: {mm_inputs}")
+        return combined_mm_item
 
     @staticmethod
     def _extract_processor_features(
-        items: List[Any], attr_name: str
+        items: List[dict], attr_name: str
     ) -> Optional[torch.Tensor]:
         """
         Helper function to concat extracted attributes from processor output.
         """
         values = [
-            getattr(item, attr_name)
+            item.get(attr_name)
             for item in items
-            if getattr(item, attr_name) is not None
+            if item.get(attr_name) is not None
+            # FIXME: Do we really need to check if the attribute is in the item?
```

Comment:
It's usually a good practice to check attribute existent when handling a dict, if you believe certain attribute always exist, I would suggest replacing dict with a dataclass to help prevent future changes breaking your assumptions.

If your concern is about code verbosity, you could probably make it more concise like this:

            value
            for item in items
            if (value := item.get(attr_name)) is not None